File: spring-graphql/src/main/java/com/example/demo/gql/DataFetchers.java
package com.example.demo.gql;

import com.example.demo.gql.types.*;
import com.example.demo.service.PostService;
import com.fasterxml.jackson.databind.ObjectMapper;
import graphql.schema.DataFetcher;
import graphql.schema.DataFetchingEnvironment;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.dataloader.DataLoader;
import org.reactivestreams.Publisher;
import org.springframework.stereotype.Component;
import org.springframework.util.StreamUtils;
import org.springframework.web.multipart.MultipartFile;

import java.nio.charset.StandardCharsets;
import java.util.List;
import java.util.UUID;
import java.util.concurrent.CompletionStage;

@RequiredArgsConstructor
@Slf4j
@Component
public class DataFetchers {
    final PostService postService;
    final ObjectMapper objectMapper;

    public DataFetcher<Post> postById() {
        return (DataFetchingEnvironment dfe) -> {
            String postId = dfe.getArgument("postId");
            return postService.getPostById(postId);
        };
    }

    public DataFetcher<List<Post>> allPosts() {
        return dfe -> postService.getAllPosts();
    }

    public DataFetcher<UUID> createPost() {
        return (DataFetchingEnvironment dfe) -> {
            var input = dfe.getArgument("createPostInput");
            var createPostInput = objectMapper.convertValue(input, CreatePostInput.class);
            return postService.createPost(createPostInput);
        };
    }

    public DataFetcher<CompletionStage<Author>> authorOfPost() {
        return (DataFetchingEnvironment dfe) -> {
            DataLoader<String, Author> dataLoader = dfe.getDataLoader("authorsLoader");
            Post post = dfe.getSource();
            return dataLoader.load(post.getAuthorId());
        };
    }

    public DataFetcher<CompletionStage<List<Comment>>> commentsOfPost() {
        return (DataFetchingEnvironment dfe) -> {
            DataLoader<String, List<Comment>> dataLoader = dfe.getDataLoader("commentsLoader");
            Post post = dfe.getSource();
            return dataLoader.load(post.getId());
        };
    }

    public DataFetcher<UUID> addComment() {
        return (DataFetchingEnvironment dfe) -> {
            var commentInputArg = dfe.getArgument("commentInput");
            var input = objectMapper.convertValue(commentInputArg, CommentInput.class);
            return this.postService.addComment(input);
        };
    }

    public DataFetcher<Publisher<Comment>> commentAdded() {
        return (DataFetchingEnvironment dfe) -> {
            log.info("connect to `commentAdded`");
            return this.postService.commentAdded();
        };
    }

    public DataFetcher<Boolean> upload() {
        return (DataFetchingEnvironment dfe) -> {
            MultipartFile file = dfe.getArgument("file");
            log.info("file name: {}", file.getName());
            log.info("content type: {}", file.getContentType());
            log.info("original file name: {}", file.getOriginalFilename());
            log.info("file content size: {}", file.getSize());
            String content = StreamUtils.copyToString(file.getInputStream(), StandardCharsets.UTF_8);
            log.info("file content : {}", content);
            return true;
        };
    }
}

File: graphql-kotlin/src/main/kotlin/com/example/demo/gql/datafetchers/DataFetchers.kt
package com.example.demo.gql.datafetchers

import com.example.demo.gql.types.Author
import com.example.demo.gql.types.Comment
import com.example.demo.gql.types.Post
import com.expediagroup.graphql.server.extensions.getValueFromDataLoader
import graphql.schema.DataFetcher
import graphql.schema.DataFetchingEnvironment
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Component
import java.util.concurrent.CompletableFuture

// lazily loading data from dataloaders
@Component("CommentListDataFetcher")
class CommentsDataFetcher : DataFetcher<CompletableFuture<List<Comment>>> {
    companion object {
        private val log = LoggerFactory.getLogger(CommentsDataFetcher::class.java)
    }

    override fun get(environment: DataFetchingEnvironment): CompletableFuture<List<Comment>> {
        val postId = environment.getSource<Post>().id
        log.debug("Fetching comments of post: $postId")
        return environment.getValueFromDataLoader(CommentsDataLoader.name, postId)
    }
}

@Component("AuthorDataFetcher")
class AuthorDataFetcher : DataFetcher<CompletableFuture<Author>> {
    companion object {
        private val log = LoggerFactory.getLogger(AuthorDataFetcher::class.java)
    }

    override fun get(environment: DataFetchingEnvironment): CompletableFuture<Author> {
        val authorId = environment.getSource<Post>().authorId
        log.debug("Fetching author of post: $authorId")
        return environment.getValueFromDataLoader(AuthorsDataLoader.name, authorId)
    }
}